### PR TITLE
Adds area-based wire layouts to airlocks.

### PR DESCRIPTION
**Pull request by Timberpoes**
## About The Pull Request

Modifies the area datum to specify an airlock wire datum for that area.

Modifies various areas to include unique departmental wire datums.

A brief overview of area wire designations from Space_Station_13_areas.dm

MAINT
	/area/maintenance/...
	
COMMAND
	/area/bridge/...
	/area/crew_quarters/heads/...
	/area/teleporter/...
	/area/gateway/...
	/area/security/nuke_storage/...
	/area/ai_monitored/nuke_storage/...

SERVICE
	/area/crew_quarters/kitchen/...
	/area/crew_quarters/bar/...
	/area/quartermaster/...
	/area/janitor/...
	/area/hydroponics/...
	
SECURITY
	/area/security/...
	/area/ai_monitored/security/armory/...

ENGINEERING
	/area/engine/...
	/area/solar/...
	/area/tcommsat/...
	
MEDBAY
	/area/medical/...
	
SCIENCE
	/area/science/...
	
AI
	/area/ai_monitored/storage/satellite/...
	/area/ai_monitored/turret_protected/...
	/area/aisat/...

PUBLIC
	/area/hallway/...
	/area/crew_quarters/...
	/area/library/...
	/area/chapel/...
	/area/lawoffice/...
	/area/construction/...
	/area/storage/...
	/area/ai_monitored/storage/eva/...
	
Public is basically the generic airlock wire assignments (in other words, these areas do not override the default wires). All other areas use generic airlock wire assignments, however it would not be difficult to add "ss13_generic" wire datum if you want the station in its entirety to have a different wirenet from anything off-station.

Of course, re-using on-station areas in off-station locations will share wire assignments.

Because wire assignments are set on airlock Init(), this means that ANY airlock constructed will have its wires automatically assigned based on the area. If necessary, I can include a mapload check in Initialise and make all doors created after round-start just have generic wires.

There should probably be a map sweep before merging to identify edge cases where maint doors are covered under departmental areas and doors that border multiple departments aren't appropriate set.

Things that come to mind are Delta's vault, which has a door assigned to the hallway area and Delta's Engi/Xenobio maint, that has maint doors covered under their respective department areas.

## Why It's Good For The Game

Hacking takes a little bit more effort than learning just one set of wires, however hacking isn't totally random. Departmental areas will share a common wire layout.

Encourages more cooperation and reliance between players when hacking. This includes sharing wire assignments for specific areas, stealing blueprints, grabbing insuls, asking engy borgos to read wire assignments for you, or figuring out another means of access that isn't hacking - Including bombs, B&E, finding weak walls to deconstruct/turn into hidden doors, finding higher level access ID cards to just walk in, etc.

## Changelog
:cl:
tweak: CentCom scientists and engineers have upgraded airlock doors throughout NanoTrasen stations. When constructed, airlock doors now automatically configure themselves based on location, adopting unique wire layouts in different station areas. Pray your CE knows how to read his blueprints, pack your marigolds and joyride a cyborg into a new age of departmental security.
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
you sneaky bastard, well done

**Comment by Winterous** [comment]
I kinda like this idea, but I feel it makes more sense to simply have harder-to-hack doors on secure areas.
They could have dead-end wires which drop bolts and temporarily depower then interacted with, they could auto-bolt when you open the panel so you need to figure out which is the bolt wire in order to hack them; just some obfuscation to make it harder to figure out, without needing to remember 5 different sets of wires like this would require.

**Comment by RaveRadbury** [comment]
Where does the vault fall in the area scheme?

**Comment by CRITAWAKETS** [comment]
add some sort of indicator on an airlock of which kind of wiring it is

**Comment by spessbro** [comment]
ok so I just wanna say props for pring something so ballsy because I can't imagine the absolute tider shitstorm this is gonna cause. But I'm wondering how this is gonna affect stuff like space ruins, bs shelters and constructed airlocks out of areas and think it would be nice to add some clarification.

**Comment by Timberpoes** [comment]
> Where does the vault fall in the area scheme?

/area/security/nuke_storage/...
/area/ai_monitored/nuke_storage/...

The above generally cover the vault, and both fall under Command right now.

> I feel it makes more sense to simply have harder-to-hack doors on secure areas

I didn't want to touch the difficulty of hacking. I agree that hacking secure areas is too easy, but I'm not minded to make hacking harder. I wanted to increase the value of other methods of breaking into an area - A number of which now have reliance on other players. The main thing is, once you've found the wire that shocks you, you've hacked every single wire on that door. You just don't pulse that wire again and you can run down and find bolts, AI, ID, etc.

Breaking in now requires a touch more preparation for an under-equipped player. It is now slightly longer for an under-prepared player. Every core department has at least one maint backdoor, many have multiple. However, just going into maint and learning the maint wires alone isn't enough. Now you have to go a step further and figure out a department's wires too - And there are non-hacking methods to do this including asking another player to find out for you, asking a cyborg to run their blueprints and give you the wires, etc.

> add some sort of indicator on an airlock of which kind of wiring it is

Actually a not-bad idea.

>  I'm wondering how this is gonna affect stuff like space ruins, bs shelters and constructed airlocks out of areas and think it would be nice to add some clarification.

I kinda clarified in the PR text, but - If the area a given map uses is one shared with the station, then it will inherit the station wires. Ideally other maps shouldn't be doing this and this is a fairly easy thing to fix.

By default - All airlocks will just have the generic airlock wires unless that airlock is in a specific area that has overridden the wire layout. As of current, the only areas that override wire layouts are those contained in Space_Station_13_areas.dm which basically contains all the areas intended for use on the station (as opposed to off station and in ruins)

**Comment by OmegaShoots** [comment]
i ded pr

**Comment by LemonInTheDark** [comment]
I wanna hear this story, how do you die to hacking.

**Comment by Timberpoes** [comment]
image
image
image
image

> add some sort of indicator on an airlock of which kind of wiring it is

So, as per the above suggestion, I've added a notice to the top. Where a set of wires comes with a `proper_name` that isn't just "Unknown", they will display the notice box stating the wire configuration. This behaviour is now active across all hackable things that use /datum/wires. As it's out of the feature scope of this PR, in a future PR wire datums can have proper_names set accordingly.

This will allow a player to see which wire layout is being used in any given airlock.

**Comment by Rohesie** [comment]
@81Denton , I believe you had plans for wires. Does this meet them?

File: code/datums/wires/_wires.dm [hunk 1]
```diff
@@ -20,33 +20,47 @@
...
-	var/list/assemblies = list() // List of attached assemblies.
-	var/randomize = 0 // If every instance of these wires should be random.
-					  // Prevents wires from showing up in station blueprints
+	/// Dictionary of colours to wire.
+	var/list/colors = list()
+	/// List of attached assemblies.
+	var/list/assemblies = list()
+
+	/// If every instance of these wires should be random. Prevents wires from showing up in station blueprints.
+	var/randomize = 0
```
Comment (Rohesie) [inline code/datums/wires/_wires.dm:42]:
```suggestion
	var/randomize = FALSE
```

File: code/datums/wires/_wires.dm [hunk 2]
```diff
@@ -20,33 +20,47 @@
...
 
 /datum/wires/New(atom/holder)
 	..()
 	if(!istype(holder, holder_type))
 		CRASH("Wire holder is not of the expected type!")
 
 	src.holder = holder
+
+	// If there is a dictionary key set, we'll want to use that. Otherwise, use the holder type.
+	var/K = dictionary_key ? dictionary_key : holder_type
```
Comment (Rohesie) [inline code/datums/wires/_wires.dm:52]:
```suggestion
	var/key = dictionary_key ? dictionary_key : holder_type
```
better var names, please

File: code/game/machinery/doors/airlock.dm [hunk 3]
```diff
@@ -1478,6 +1478,16 @@
...
 
+/**
+  *	Generates the airlock's wire layout based on the current area the airlock resides in.
+  *
+  * Returns a new /datum/wires/ with the appropriate wire layout based on the airlock_wires
+  * of the area the airlock is in.
+  */
+/obj/machinery/door/airlock/proc/set_wires()
+	var/area/A = get_area(src)
+	return new A.airlock_wires(src)
```
Comment (Rohesie) [inline code/game/machinery/doors/airlock.dm:1489]:
```suggestion
	var/area/source_area = get_area(src)
	return new source_area.airlock_wires(src)
```

**Comment by Rohesie** [comment]
I'll merge this in a day or two if no maintainer speaks against it.

**Comment by 81Denton** [comment]
@Rohesie Sorry about that, your ping got lost in my notifications. This is fine.

**Pull request merged**